p4-16/psa/psa.p4

```diff
@@ -65,6 +65,28 @@ const   CloneSessionId_t PSA_CLONE_SESSION_TO_CPU = unspecified;
 
 #endif
 
+// BEGIN:Type_defns2
+
+/* Note: All of the widths for types with `InHeader` in their name are
+ * intended only to carry values of the corresponding types in packet
+ * headers between a controller and a PSA device.  The widths are
+ * intended to be large enough for all PSA devices, so that a
+ * controller can fill in headers with these fields in a common way
+ * for all PSA devices.
+ *
+ * All widths must be a multiple of 8, so that any subset of these
+ * fields may be used in a single P4 header definition, even on P4
+ * implementations that restrict headers to contain fields with a
+ * total length that is a multiple of 8 bits. */
+typedef bit<16> PortIdInHeader_t;
+typedef bit<32> MulticastGroupInHeader_t;
+typedef bit<16> CloneSessionIdInHeader_t;
+typedef bit<8>  ClassOfServiceInHeader_t;
+typedef bit<16> PacketLengthInHeader_t;
+typedef bit<16> EgressInstanceInHeader_t;
+typedef bit<64> TimestampInHeader_t;
```

Comment:
Should we explicitly state that we do not need translation annotations for these other types?

Reply:
@vgurevich @antoninbas @samar-abdi Any thoughts on whether any of these other types would need conversion of values?

Can the control plane API just use whatever multicast group numbers and egress instance numbers it wants?  I assume different implementations will have different numbers of these things that they support, and some might have more restricted numerical ranges than others.  Does a P4Runtime API controller have a way to discover such restrictions on these values for different PSA devices in a heterogenous network?

Do we expect ClassOfService_t values to be numerically consecutive in a range [0, N] for all PSA devices, and should say that in PSA?

Same question for CloneSessionId values.

I _think_ we are safe from such issues for PacketLength_t and Timestamp_t values, i.e. some implementations might have different maximum PacketLength_t values they actually support, but all of them will be [my_min, my_max] for packet length support for some my_min, my_max integers, and timestamps should be all X-bit-wide values for the value of X supported by the implementation.

Reply:
Very likely class of service will need translation. It has the same issue as ports. Different controllers may have different notions of classes of service and their own numbering scheme for them. The switch config handler on the target will map the controller's class of service to the ASIC's. Other metadata values are either programmed by the controller (mcast_group/egress instance) or have universally agreed values (packet_length/timestamp) IMO.

Reply:
So we should explicitly document that type ClassOfService_t fields will be translated between controller and P4 data plane programs, as PortId_t fields will be, but _no others will_ ?

Do we expect, and should we perhaps document, that for all of the following types, the data plane and the control plane will both support _all_ numerical values in the range [0, N] for whatever maximum value of N is supported by the PSA device (which can vary from one PSA implementation to another)?

```
typedef bit<unspecified> MulticastGroup_t;
typedef bit<unspecified> CloneSessionId_t;
typedef bit<unspecified> EgressInstance_t;
```

Reply:
Yes for documenting translation of ClassOfService_t field values. The P4Runtime annotation will be `cos_translation("controller_bitwidth:32")`.

As for your second point, I suggest : "The values of multicast_group, clone_session and egress_instance are completely in the purview of the controller, and are limited only by the bitwidth of the respective types on the PSA device(s) being controlled by the controller."

I believe it is implicit that if a psa.p4 has bit<7> multicast_group_id, it will allow all values from 0 to 127.

Reply:
I know it is common for things in computers to have sizes that are powers of 2, but it sometimes happens that people implement SRAMs with a number of entries that aren't powers of 2 (or a collection of them whose total number of entries is not a power of 2, more likely, but have a common address space spanning all of them).

I would hate for a controller to have an easily avoided bug if this turns out to be the case, because it made such power-of-2 size assumption.